compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/analyze.rs

```diff
@@ -114,25 +112,40 @@ pub fn analyze(
     })
 }
 
-fn no_input_var_in_out_var(
+fn check_composability(
     dag: &unparametrized::OperationDag,
     symbolic_variances: &[Vec<SymbolicVariance>],
     nb_partitions: usize,
-) -> bool {
-    // let a = dag.get_output_index()
-    //     .iter()
-    //     .flat_map(|index| symbolic_variances[*index].iter()).collect::<Vec<_>>();
-    // println!("symb_variances: {:?}", a);
-
-    dag.get_output_index()
+) -> Result<()> {
+    // If the circuit only contains inputs, then it is composable.
+    let only_inputs = dag
+        .operators
         .iter()
-        .flat_map(|index| symbolic_variances[*index].iter())
-        .all(|sym_var| {
-            (0..nb_partitions).all(|partition| {
-                let coeff = sym_var.coeff_input(partition);
-                coeff == 0.0f64 || coeff.is_nan()
-            })
-        })
+        .all(|node| matches!(node, Operator::Input { .. }));
+    if only_inputs {
+        return Ok(());
+    }
+
+    // If the circuit outputs are free from input variances, it means that every outputs are
+    // refreshed, and the function can be composed.
+    let in_var_in_out_var = dag
```

Comment:
IMO 2 for loops would be simpler here:

```rust
let mut ... = None;
for index in dag.get... {
   let sym_var = ...
   for partition in ... {
        if ... {
            ... = ...;
            break;
        }
   }
 }
}
```

Reply:
I prefer combinators TBH. Even (more) when the control flow is not straightforward. The region of code you have to look at to understand each line of code is often very reduced. Which is not the case when using loops and mutable variables. Plus, even if it is not very relevant here, the generated code is often faster, because the dependencies between intermediate state values are better expressed.